src/android/Sketch.java

```diff
@@ -265,44 +272,63 @@ public void run() {
+        if (drawingPath == null || drawingPath.length() == 0) {
             LOG.e(TAG, "Failed to read sketch result from activity");
             this.callbackContext.error("Failed to read sketch result from activity");
             return;
         }
 
         try {
-            String ext = "";
+            if (destinationType == DestinationType.DATA_URL) {
+                String dataenc = "";
+                byte[] drawingData = null;
 
-            if (encodingType == EncodingType.JPEG) {
-                ext = "jpeg";
-            } else if (encodingType == EncodingType.PNG) {
-                ext = "png";
-            }
+                if (encodingType == EncodingType.JPEG) {
+                    dataenc = "jpeg";
+                } else if (encodingType == EncodingType.PNG) {
+                    dataenc = "png";
+                }
 
-            if (destinationType == DestinationType.DATA_URL) {
-                output = "data:image/" + ext + ";base64," + Base64.encodeToString(drawingData, Base64.DEFAULT);
-            } else if (destinationType == DestinationType.FILE_URI) {
-                // Save the drawing to the app's cache dir
-                String fileName = String.format("sketch-%s.%s", UUID.randomUUID(), ext);
-                File filePath = new File(this.cordova.getActivity().getCacheDir(), fileName);
+                LOG.d(TAG, "Reading temp file from: " + drawingPath);
+                // decode the file and convert to byte array
+                Bitmap bm = BitmapFactory.decodeFile(drawingPath);
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+                // write the image data to output stream using the right encoding type
+                if (encodingType == EncodingType.PNG) {
+                    bm.compress(Bitmap.CompressFormat.PNG, 100, baos);  
+                } else {
+                    bm.compress(Bitmap.CompressFormat.JPEG, 100, baos);  
+                }
+ 
+                // convert stream to array data
+                drawingData = baos.toByteArray();
+
+                // error out if we didn't successfully read back the temp file
+                if (drawingData == null || drawingData.length == 0) {
+                    LOG.e(TAG, "Failed to read sketch result from activity");
+                    this.callbackContext.error("Failed to read sketch result from activity");
+                    return;
+                }
 
-                FileOutputStream fos = new FileOutputStream(filePath);
-                fos.write(drawingData);
-                fos.close();
+                output = "data:image/" + dataenc + ";base64," + Base64.encodeToString(drawingData, Base64.DEFAULT);
+            } else if (destinationType == DestinationType.FILE_URI) {
+                // Get the filename based on the absolute path we received
+                String fileName = drawingPath.substring(drawingPath.lastIndexOf("/")+1);
```

Comment:
Could we possible use http://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html to get the file name. e.g.

```java
Path p = Paths.get(drawingPath);
String fileName = p.getFileName().toString();
```